Add samplesheet examples (#82)
nextflow-io/nf-validation: docs/samplesheets/examples.md

# Sample sheet channel manipulation examples

Each item in the channel emitted by `.fromSamplesheet()` is a flat tuple, corresponding with each row of the samplesheet. Each item will be composed of a meta value (if present) and any additional elements from columns in the samplesheet, e.g.:

```csv
sample,fastq_1,fastq_2,bed
sample1,fastq1.R1.fq.gz,fastq1.R2.fq.gz,sample1.bed
sample2,fastq2.R1.fq.gz,fastq2.R2.fq.gz,
```

Might create the channel:

```nextflow

> tuple val(meta), path(fastq_1), path(fastq_2), path(bed)

Reviewer: I'm not entirely convinced this is the cleanest and most clear way of representing the channel structure, maybe something like `[meta, fastq_1.fastq, fastq_2.fastq, sample1.bed]` will be more clear for beginners?

Reviewer: It was to emulate a filename, but yeah this is just a matter of representation. I'm not sure which way is the best or most clear to beginners. (although both should be fine I think)

# Sample sheet channel manipulation examples

Each item in the channel emitted by `.fromSamplesheet()` is a flat tuple, corresponding with each row of the samplesheet. Each item will be composed of a meta value (if present) and any additional elements from columns in the samplesheet, e.g.:

```csv
sample,fastq_1,fastq_2,bed
sample1,fastq1.R1.fq.gz,fastq1.R2.fq.gz,sample1.bed
sample2,fastq2.R1.fq.gz,fastq2.R2.fq.gz,
```

Might create a channel where each element consists of 4 items, a map value followed by three files:

```groovy
// Columns:
[ val([ sample: sample ]), file(fastq1), file(fastq2), file(bed) ]

// Resulting in:
[ [ id: "sample" ], fastq1.R1.fq.gz, fastq1.R2.fq.gz, sample1.bed]

> [ [ id: "sample2" ], fastq2.R1.fq.gz, fastq2.R2.fq.gz, null ] // A missing value from the samplesheet is null

Reviewer: ```suggestion
[ [ id: "sample2" ], fastq2.R1.fq.gz, fastq2.R2.fq.gz, [] ] // A missing value from the samplesheet is `[]`
```

Files will default to [] when they aren't given (null will cause errors in processes where the files are used)

### Glossary

- A channel is the Nextflow object, referenced in the code
- An item is each thing passing through the channel, equivalent to one row in the samplesheet
- An element is each thing in the item, e.g., the meta value, fastq_1 etc. It may be a file or value

## Default mode

Each item in the channel emitted by `.fromSamplesheet()` is a flat tuple, corresponding with each row of the samplesheet. Each item will be composed of a meta value (if present) and any additional elements from columns in the samplesheet, e.g.:

```csv
sample,fastq_1,fastq_2,bed
sample1,fastq1.R1.fq.gz,fastq1.R2.fq.gz,sample1.bed
sample2,fastq2.R1.fq.gz,fastq2.R2.fq.gz,
```

Might create a channel where each element consists of 4 items, a map value followed by three files:

```groovy
// Columns:
[ val([ sample: sample ]), file(fastq1), file(fastq2), file(bed) ]

// Resulting in:
[ [ id: "sample" ], fastq1.R1.fq.gz, fastq1.R2.fq.gz, sample1.bed]
[ [ id: "sample2" ], fastq2.R1.fq.gz, fastq2.R2.fq.gz, [] ] // A missing value from the samplesheet is an empty list
```

> This could be used in the Nextflow process as:

Reviewer 2: This sentence sounds a bit confusing to me, as I would expect seeing the use of the channel instead of the process input declaration. Could we rephrase with something like "This channel can be used as input of a process where the input declaration is:"

```nextflow
tuple val(meta), path(fastq_1), path(fastq_2), path(bed)
```

It may be necessary to manipulate this channel to fit your process inputs. For more documentation, check out the [Nextflow operator docs](https://www.nextflow.io/docs/latest/operator.html), however here are some common use cases with `.fromSamplesheet()`.

## Changing the structure of channel items

Each item in the channel will be a flat tuple, but some processes will use multiple files as a list in their input channel, this is common in nf-core modules. For example, consider the following input declaration in a process, where FASTQ could be > 1 file:

```nextflow
process ZCAT_FASTQS {
    input:
        tuple val(meta), path(fastq)

    """
    zcat $fastq
    """
}
```

> The output of `.fromSamplesheet()` is the following by default:

Reviewer 2: Similar to my previous comment, I would try to differenciate between the channel and the input declaration. For example
```suggestion
The output of `.fromSamplesheet()` can be used by default with a process with the following input declaration:
```

```nextflow

> val(meta, path(fastq_1), path(fastq_2)

Reviewer 2: ```suggestion
val(meta), path(fastq_1), path(fastq_2)
```

```nextflow
Channel.fromSamplesheet("input")
    .map { meta, fastq_1, fastq_2, bed -> tuple(meta, fastq_1, fastq_2) }
    .set { input }

input.view() // Channel has 3 elements: meta, fastq_1, fastq_2
```

In this way you can drop items from a channel.

## Separating channel items

We could perform this twice to create one channel containing the FASTQs and one containing the BED files, however Nextflow has a native operator to separate channels called [`.multiMap()`](https://www.nextflow.io/docs/latest/operator.html#multimap). Here, we separate the FASTQs and BEDs into two separate channels using `multiMap`. Note, the channels are both contained in `input` and accessed as an attribute using dot notation:

```nextflow
Channel.fromSamplesheet("input")
    .multiMap { meta, fastq_1, fastq_2, bed ->
        fastq: tuple(meta, fastq_1, fastq_2)
        bed:   tuple(meta, bed)
    }
    .set { input }
```

The channel has two attributes, `fastq` and `bed`, which can be accessed separately.

```nextflow
input.fastq.view() // Channel has 3 elements: meta, fastq_1, fastq_2
input.bed.view()   // Channel has 2 elements: meta, bed
```

> Importantly, `multiMap` apples to every item in the channel and returns an item to both channels for every input, i.e. `input`, `input.fastq` and `input.bed` all contain the same number of items, however each item will be different.

Reviewer 2: ```suggestion
Importantly, `multiMap` applyes to every item in the channel and returns an item to both channels for every input, i.e. `input`, `input.fastq` and `input.bed` all contain the same number of items, however each item will be different.
```
